native/xqlitenif/src/r2d2.rs

```diff
@@ -0,0 +1,445 @@
+}
+
+fn create_pool(path: &str) -> Result<u64, r2d2::Error> {
+    let manager = SqliteConnectionManager::file(path);
+    let pool = Pool::builder().build(manager)?;
+    let id = NEXT_ID.fetch_add(1, Ordering::SeqCst);
+    pools().insert(id, pool);
+    Ok(id)
+}
+
+fn get_pool(id: u64) -> Option<XqlitePool> {
+    pools().get(&id).map(|x| x.value().clone())
+}
+
+fn remove_pool(id: u64) -> bool {
+    pools().remove(&id).is_some()
+}
+
+fn xqlite_open(path: &str) -> Result<ResourceArc<XqliteConn>, XqliteError> {
+    let path_clone = path.to_string();
+    let pool_id = create_pool(path)
+        .map_err(|e| XqliteError::CannotOpenDatabase(path_clone, e.to_string()))?;
+
+    let conn_handle = XqliteConn(pool_id);
+    let resource = ResourceArc::new(conn_handle);
+    Ok(resource)
+}
+
+fn xqlite_exec<'a>(
+    env: Env<'a>,
+    handle: &XqliteConn,
+    sql: &str,
+    params_term: Term<'a>,
+) -> Result<Vec<Vec<XqliteVal>>, XqliteError> {
+    let pool = get_pool(handle.0).ok_or(XqliteError::ConnectionNotFound(*handle))?;
+    let conn = pool.get()?;
+    let named_params_vec = decode_keyword_params(env, params_term)?;
+
+    let sql_string = sql.to_string();
+    let mut stmt = conn
+        .prepare(sql)
+        .map_err(|e| XqliteError::CannotPrepareStatement(sql_string, e.to_string()))?;
+
+    let column_count = stmt.column_count();
+
+    let params_for_rusqlite: Vec<(&str, &dyn ToSql)> = named_params_vec
+        .iter()
+        .map(|(k, v)| (k.as_str(), v as &dyn ToSql))
+        .collect();
+
+    let mut rows = stmt.query(params_for_rusqlite.as_slice())?;
+
+    let mut results: Vec<Vec<XqliteVal>> = Vec::new();
+
+    while let Some(row) = rows
+        .next()
+        .map_err(|e| XqliteError::CannotFetchRow(e.to_string()))?
+    {
+        let mut row_values: Vec<XqliteVal> = Vec::with_capacity(column_count);
+        for i in 0..column_count {
+            let value = row
+                .get::<usize, Value>(i)
+                .map_err(|e| XqliteError::CannotReadColumn(i, e.to_string()))?;
+            row_values.push(XqliteVal(value));
+        }
+        results.push(row_values);
+    }
+
+    Ok(results)
+}
+
+fn xqlite_pragma_write(handle: &XqliteConn, pragma_sql: &str) -> Result<usize, XqliteError> {
+    let pool = get_pool(handle.0).ok_or(XqliteError::ConnectionNotFound(*handle))?;
+    let conn = pool.get()?;
+
+    conn.execute(pragma_sql, [])
+        .map_err(|e| XqliteError::CannotExecutePragma {
+            pragma: pragma_sql.to_string(),
+            reason: e.to_string(),
+        })
+}
+
+fn xqlite_close(handle: &XqliteConn) -> Result<(), XqliteError> {
+    if remove_pool(handle.0) {
+        Ok(())
+    } else {
+        Err(XqliteError::ConnectionNotFound(*handle))
+    }
+}
+
+#[rustler::nif(schedule = "DirtyIo")]
+fn raw_open(env: Env, path: String) -> Term {
+    match xqlite_open(&path) {
+        Ok(resource) => (ok(), resource).encode(env),
+        Err(err) => (error(), err).encode(env),
```

Comment:
I don't think this step is necessary, we implement an `Encoder` for `Result` that does exactly this.

Reply:
So I should remove the `env: Env` argument and make the return value be `NifResult<...>` instead?

Guess I got a bit carried away here because I had a case where I only wanted to return `:ok` and nothing else and figured I'll just unify how I return stuff from all my NIF functions here. 🤔

Reply:
Hm, just tested this by removing the current `raw_open` and renaming `xqlite_open` -> `raw_open`. Worked without a hitch. This is what happens when you code in a frenzy. 🤦 I'll remove this unnecessary indirection.

Reply:
Done, the wrappers are gone: https://github.com/dimitarvp/xqlite/pull/9/commits/022f2f171ccb7d6d7f12555e361a4c20cee70f3b